### PR TITLE
TargetLocker.Refresh and .Unlock API changes, clean up tests

### PR DESCRIPTION
**Pull request by rojer9-fb**
Change the semantics of TargetLocker.Refresh to require that owner holds
locks for the requested targets (having an expired lock for the same job
is also permitted).

This is necessary for job resumption: when resuming a paused job we need
to be sure that targets have not in the mean time been assigned to other
jobs and thus had their state interfered with.

Some drive-by improvements:

 * Added TargetLocker.Close() to clean up locker instances. Used in tests.
 * Parametrise lock lease length, expose as `-targetLockDuration`
   argument and increase the default to 10 minutes.
 * Unify testing for TargetLocker implementations: use the same suite of
   tests for both inmemory and dblocked implementations, resolved subtle
   differences exposed by this.
 * Passing empty target list to the locker API is now an error (used to
   be allowed by dblocked but not inmemory).
 * Trying to unlock a target that was not locked by the specified job is
   an now an error.
 * Wire clock.Clock through job manager and runner API, for use in tests.
   Actual use is a TODO (turned out to be non-trivial)
 * Speed up job manager shutdown (speeds up tests)

### CODE REVIEW

**Review by tfg13 — COMMENTED** [review]
I'll review this tomorrow, but I am 99% sure internal plugins call this with zero targets and expect it to not error

**Comment by rojer9-fb** [comment]
i doubt it because inmemory lock [fails requests with no targets](https://github.com/facebookincubator/contest/blob/984dca4f91fb995ad3ec073d8df5db0c058a0252/plugins/targetlocker/inmemory/inmemory.go#L60-L62).

**Comment by tfg13** [comment]
Ha you are right (D25060297)

**Comment by rojer9-fb** [comment]
rebased

File: plugins/targetlocker/dblocker/dblocker.go [hunk 1]
```diff
@@ -209,20 +217,47 @@ func (d *DBLocker) handleLock(ctx xcontext.Context, jobID int64, targets []strin
...
 		return fmt.Errorf("unable to unlock targets %v, owner %d: %w", targets, jobID, err)
 	}
-	return nil
+	return tx.Commit()
 }
 
 func validateTargets(targets []*target.Target) error {
+	if len(targets) == 0 {
+		return fmt.Errorf("no target specified")
+	}
```
Comment (tfg13) [inline plugins/targetlocker/dblocker/dblocker.go:260]: You were right about `Lock`, but `Unlock` might be called in (very rare) error paths with zero targets, this simplifies target managers. (https://fburl.com/diffusion/evshb51n)

File: plugins/targetlocker/dblocker/dblocker.go [hunk 2]
```diff
@@ -240,11 +275,8 @@ func (d *DBLocker) Lock(ctx xcontext.Context, jobID types.JobID, duration time.D
 	if err := validateTargets(targets); err != nil {
 		return fmt.Errorf("invalid lock request: %w", err)
 	}
-	ctx.Debugf("Requested to lock %d targets for job ID %d: %v", len(targets), jobID, targets)
-	if len(targets) == 0 {
-		return nil
-	}
-	_, err := d.handleLock(ctx, int64(jobID), targetIDList(targets), uint(len(targets)), duration, false)
+	_, err := d.handleLock(ctx, int64(jobID), targetIDList(targets), uint(len(targets)), duration, false /* requireLocked */, false /* allowConflicts */)
+	ctx.Debugf("Lock %d targets for job %d for %s: %v", len(targets), jobID, duration, err)
```
Comment (tfg13) [inline plugins/targetlocker/dblocker/dblocker.go:279]: we can remove the jobID now, it is in the contextual logger (but can do in another PR)

File: cmds/contest/main.go [hunk 3]
```diff
@@ -31,14 +33,15 @@ import (
...
+	flagServerID           = flag.String("serverID", "", "Set a static server ID, e.g. the host name or another unique identifier. If unset, will use the listener's default")
+	flagProcessTimeout     = flag.Duration("processTimeout", api.DefaultEventTimeout, "API request processing timeout")
+	flagTargetLocker       = flag.String("targetLocker", inmemory.Name, "Target locker implementation to use")
+	flagInstanceTag        = flag.String("instanceTag", "", "A tag for this instance. Server will only operate on jobs with this tag and will add this tag to the jobs it creates.")
+	flagLogLevel           = flag.String("logLevel", "debug", "A log level, possible values: debug, info, warning, error, panic, fatal")
+	flagPauseTimeout       = flag.Duration("pauseTimeout", 0, "SIGINT/SIGTERM shutdown timeout (seconds), after which pause will be escalated to cancellaton; -1 - no escalation, 0 - do not pause, cancel immediately")
+	flagResumeJobs         = flag.Bool("resumeJobs", false, "Attempt to resume paused jobs")
+	flagTargetLockDuration = flag.Duration("targetLockDuration", config.DefaultTargetLockDuration, "The amount of time target lock is extended by while the job is running.")
```
[the remark this answers is not in the document]
Comment (rojer9-fb) [reply cmds/contest/main.go:44]: i assume you meant job. done.

File: pkg/runner/job_runner.go [hunk 4]
```diff
@@ -228,7 +245,7 @@ func (jr *JobRunner) Run(ctx xcontext.Context, j *job.Job, resumeState *job.Paus
 					close(stopRefresh)
 					<-stoppedRefresh
 					// Refresh one last time.
-					if err2 := tl.RefreshLocks(runCtx, j.ID, targets); err2 != nil {
+					if err2 := tl.Lock(runCtx, j.ID, jr.targetLockDuration, targets); err2 != nil {
```
[the remark this answers is not in the document]
Comment (rojer9-fb) [reply pkg/runner/job_runner.go:248]: oh, good catch. this is a leftover from early version of the change where i actually thought we didn't need RefreshLocks and could do with Lock everywhere - since before this change Refresh was just Lock with a different (and constant) timeout. when i went to parametrize the timeout used by refresh i realized that it was exactly the same as Lock and wanted to get rid of it. but then, after more thought i realized that no, we needed a separate operation with different semantics.

File: plugins/targetlocker/dblocker/dblocker.go [hunk 5]
```diff
@@ -209,20 +217,47 @@ func (d *DBLocker) handleLock(ctx xcontext.Context, jobID int64, targets []strin
...
+	if err != nil {
+		return err
+	}
+	for _, t := range targets {
+		l, ok := locks[t]
+		if !ok {
+			return fmt.Errorf("target %q is not locked", t)
+		}
+		if l.jobID != jobID {
+			return fmt.Errorf("unlock request: target %q is locked by %q, not by %q", t, l.jobID, jobID)
```
[the remark this answers is not in the document]
Comment (rojer9-fb) [reply plugins/targetlocker/dblocker/dblocker.go:240]: updated. i think previous behavior was too lax. we need to be loud when targets are stolen from the job, or we'll have difficult to trace failures.

File: plugins/targetlocker/dblocker/dblocker.go (hunk 2 above)
Comment (rojer9-fb) [reply plugins/targetlocker/dblocker/dblocker.go:279]: done

File: plugins/targetlocker/dblocker/dblocker.go (hunk 1 above)
Comment (rojer9-fb) [reply plugins/targetlocker/dblocker/dblocker.go:260]: ok, as discussed, i modified the API to allow empty target lists for all methods.

File: plugins/targetlocker/dblocker/dblocker.go [hunk 6]
```diff
@@ -138,11 +136,21 @@ func (d *DBLocker) handleLock(ctx xcontext.Context, jobID int64, targets []strin
 	for _, targetID := range targets {
 		lock, ok := locks[targetID]
 		switch {
-		case !ok:
+		case !ok: // nonexistent lock
+			if requireLocked {
+				return nil, fmt.Errorf("target %q must be locked but isn't", targetID)
+			}
```
[the remark this answers is not in the document]
Comment (rojer9-fb) [reply plugins/targetlocker/dblocker/dblocker.go:142]: fair enough, done

**Comment by rojer9-fb** [comment]
PTAL

**Review by tfg13 — APPROVED** [review]
